bpo-36871: Handle spec errors in assert_has_calls

sfreilich opened the pull request:
The fix in PR #13261 handled the underlying issue about the spec for specific methods not being applied correctly, but it didn't fix the issue that was causing the misleading error message.

The code currently grabs a list of responses from _call_matcher (which may include exceptions). But it doesn't reach inside the list when checking if the result is an exception. This results in a misleading error message when one of the provided calls does not match the spec.

<!-- issue-number: [bpo-36871](https://bugs.python.org/issue36871) -->
https://bugs.python.org/issue36871
<!-- /issue-number -->


Automerge-Triggered-By: @gpshead

sfreilich committed:
bpo-36871: Handle spec errors in assert_has_calls

The fix in PR 13261 handled the underlying issue about the spec for specific methods not being applied correctly, but it didn't fix the issue that was causing the misleading error message.

The code currently grabs a list of responses from _call_matcher (which may include exceptions). But it doesn't reach inside the list when checking if the result is an exception. This results in a misleading error message when one of the provided calls does not match the spec.

sfreilich committed:
Fix error-handling bug in assert_has_calls

As well as assert_has_awaits, where I misplaced that fix previously.
Add tests for both fixes.

tirkarthi commented on Lib/unittest/mock.py:
```diff
@@ -931,7 +931,7 @@ def assert_has_calls(self, calls, any_order=False):
         they must all appear in `mock_calls`."""
         expected = [self._call_matcher(c) for c in calls]
-        cause = expected if isinstance(expected, Exception) else None
+        cause = next((e for e in expected if isinstance(e, Exception)), None)
```
Good change. Thanks. One suggestion is that this will only show the first exception on a series of exceptions. Thoughts on below?

```python
def f(): pass

mock = Mock(spec=f)
mock.assert_has_calls([call(a='wrong'), call(1)])
```

```
TypeError: got an unexpected keyword argument 'a'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/tmp/foo.py", line 15, in <module>
    mock.assert_has_calls([call(a='wrong'), call(1)])
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/unittest/mock.py", line 939, in assert_has_calls
    raise AssertionError(
AssertionError: Calls not found.
Expected: [call(a='wrong'), call(1)]
```

Something like `cause = [e for e in expected if isinstance(e, Exception)] or None` would show multiple exceptions but this would mean it should be an `Exception` and we can't get the type. 

```
Exception: [TypeError("got an unexpected keyword argument 'a'"), TypeError('too many positional arguments')]

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/tmp/foo.py", line 15, in <module>
    mock.assert_has_calls([call(a='wrong'), call(1)])
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/unittest/mock.py", line 939, in assert_has_calls
    raise AssertionError(
AssertionError: Calls not found.
Expected: [call(a='wrong'), call(1)]
```

lisroach replied:
@1st1 here's another example of where MultiError could come in handy. 

I do like the showing of the multiple arguments, it's not something we often see in Python but in the case of testing the more exceptions you can find the better it seems to me.

sfreilich replied:
I feel like I'm missing some context re MultiError, what is that?

I'm somewhat surprised putting a list of exceptions in the from clause works, the [documentation](https://docs.python.org/3.9/reference/simple_stmts.html#the-raise-statement) implies that's not correct:

> The from clause is used for exception chaining: if given, the second expression must be another exception class or instance

Could always put more information in the text of the message, though? (Instead of or in addition to setting the cause to the first exception encountered?)

tirkarthi replied:
I hope there will be a PEP soon for the idea. I guess this trio issue helps and the closest reference I can find with google : https://github.com/python-trio/trio/issues/611

sfreilich replied:
I added a change to improve the error messages to include all the exceptions encountered while processing the expected calls, while still setting the cause of the AssertionError to the first exception.

sfreilich committed:
Include all exceptions in error message

Include all exceptions deferred when binding expected calls with the
spec in the error messages generated by assert_has_calls and
assert_has_awaits. To make the origin of each exception clear, the
exceptions are presented in a list parallel to the expected calls, with
None at positions where processing the call generated no exception.

The first exception encountered is still recorded as the cause of the
AssertionError, since that's the exception which would have been raised
had the handling of those exceptions not been deferred.

sfreilich committed:
Fix whitespace lint

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

sfreilich committed:
Line breaks

(Thought the blurb-it tool would handle that.)

tirkarthi commented on Lib/unittest/mock.py:
```diff
@@ -931,13 +931,21 @@ def assert_has_calls(self, calls, any_order=False):
-                    % (_CallList(calls), self._calls_repr(prefix="Actual"))
+                    f'{problem}\n'
+                    f'Expected: {_CallList(calls)}\n'
+                    f'Actual: {self._calls_repr(prefix="Actual")}'
```
This causes a difference in error message with Actual line printed twice. Using `self._calls_repr` alone takes care of it. `self._calls_repr` already prepends a newline so please remove the "\n" in `f'Expected: {_CallList(calls)}\n'` to avoid an extra new line.

```python
from unittest.mock import *

def f(): pass

m = Mock(spec=f)
m()
m.assert_has_calls([call(), call('wrong')])
```

```
➜  cpython git:(master) ✗ ./python.exe /tmp/foo.py
TypeError: too many positional arguments

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/tmp/foo.py", line 7, in <module>
    m.assert_has_calls([call(), call('wrong')])
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/unittest/mock.py", line 945, in assert_has_calls
    raise AssertionError(
AssertionError: Error processing expected calls.
Errors: [None, TypeError('too many positional arguments')]
Expected: [call(), call('wrong')]
Actual:
Actual: [call()].
```

```suggestion
                    f'{self._calls_repr(prefix="Actual")}'
```

gpshead replied:
whooops. thanks!  i missed that. :)  follow up with a new PR with that little change.  looks like we need to do the backport manually and can lump these two changes together for that anyways.

sfreilich replied:
Good catch, sorry. I was making this more similar to cleaner logic in assert_has_awaits, and I missed that repetition.
